keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -378,6 +384,201 @@ class SqlArtifactRepository(
+    version: String,
+    targetEnvironment: String,
+    force: Boolean
+  ): Boolean {
+
+    val (envUid, artUid) = sqlRetry.withRetry(READ) {
+      Pair(
+        deliveryConfig.getUidStringFor(
+          deliveryConfig.environmentNamed(targetEnvironment)),
+        artifact.uidString)
+    }
+
+    val isPinned = sqlRetry.withRetry(READ) {
+      jooq
+        .fetchExists(
+          ENVIRONMENT_ARTIFACT_PIN,
+          ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(envUid)
+            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID.eq(artUid))
+            .and(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION.eq(version)))
+    }
+
+    if (isPinned) {
+      log.warn(
+        "Pinned artifact version cannot be vetoed: " +
+          "deliveryConfig=${deliveryConfig.name}, " +
+          "environment=$targetEnvironment, " +
+          "artifactVersion=$version")
+      return false
+    }
+
+    return jooq
+      .select(
+        ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS,
+        ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_REFERENCE)
+      .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+      .where(
+        ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(envUid),
+        ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artUid),
+        ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(version))
+      .fetchOne { (status, reference) ->
+        if (reference != null && !force) {
+          log.warn(
+            "Not vetoing artifact version as it appears to have already been an automated rollback target: " +
+              "deliveryConfig=${deliveryConfig.name}, " +
+              "environment=$targetEnvironment, " +
+              "artifactVersion=$version, " +
+              "priorVersionReference=$reference")
+          return@fetchOne false
+        }
+
+        val prior = priorVersionDeployedIn(envUid, artUid, version)
+
+        sqlRetry.withRetry(WRITE) {
+          jooq.transaction { config ->
+            val txn = DSL.using(config)
+            txn.update(ENVIRONMENT_ARTIFACT_VERSIONS)
+              .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS, VETOED.name)
+              .set(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_REFERENCE, prior ?: version)
+              .where(
+                ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(envUid),
+                ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artUid),
+                ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(version)
+              )
+              .execute()
+
+            /**
+             * If there's a previously deployed version in [targetEnvironment], set `promotion_reference`
+             * to the version that's currently being vetoed. If that version also fails to fully deploy,
+             * this is used to short-circuit further automated vetoes. We want to avoid a cloud provider
+             * or other issue unrelated to an artifact version triggering continual automated rollbacks
+             * thru all previously deployed versions.
```

Comment:
Calling attention to this comment explaining the use of `ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS`, a new column added by this PR and used to prevent recursively vetoing prior artifact versions.